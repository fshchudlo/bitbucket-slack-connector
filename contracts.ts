import * as slack from "@slack/web-api";
// see https://confluence.atlassian.com/bitbucketserver0816/event-payload-1333334207.html#Eventpayload-pullrequest

export declare interface UserPayload {
    name: string;
    emailAddress: string;
}
export declare interface PullRequestPayload {
    id: number;
    title: string;
    author: { user: UserPayload };
    reviewers: Array<{ user: UserPayload }>;
    links: { self: Array<string> };
    toRef: {
        displayId: string;
        repository: {
            slug: string;
            project: { key: string; name: string };
        };
    };
}
export declare interface PullRequestCreatedPayload {
    eventKey: string;
    date: string;
    pullRequest: PullRequestPayload;
}

export declare interface PullRequestMergedDeclinedDeletedPayload {
    eventKey: string;
    date: string;
    actor: UserPayload;
    pullRequest: PullRequestPayload;
}

export declare interface SlackGateway {
    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse>;
    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse>;
    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse>;
    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse>;
    closeChannel(options: slack.ConversationsCloseArguments): Promise<slack.ConversationsCloseResponse>;
    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse>;
}
