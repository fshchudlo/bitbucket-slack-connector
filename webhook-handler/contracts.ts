// see https://confluence.atlassian.com/bitbucketserver0816/event-payload-1333334207.html#Eventpayload-pullrequest
export declare interface PullRequestBasicPayload {
    eventKey: string;
    date: string;
    actor: UserPayload;
    pullRequest: PullRequestPayload;
}

export declare interface PullRequestCommentAddedPayload extends PullRequestBasicPayload {
    comment: {
        id: number;
        text: string;
        author: UserPayload;
    };
}

export declare interface UserPayload {
    displayName: string;
    emailAddress: string;
}

export declare interface PullRequestPayload {
    id: number;
    title: string;
    author: { user: UserPayload };
    reviewers: Array<{ user: UserPayload }>;
    links: { self: Array<{ href: string }> };
    toRef: {
        displayId: string;
        repository: {
            slug: string;
            project: { key: string; name: string };
        };
    };
}

