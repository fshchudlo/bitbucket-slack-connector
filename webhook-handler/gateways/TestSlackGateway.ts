import * as slack from "@slack/web-api";

import { SlackGateway } from "./SlackGateway";

const channelId = "12345";
export default class TestSlackGateway implements SlackGateway {
    snapshot: {
        createdChannels: slack.ConversationsCreateArguments[];
        setChannelTopics: slack.ConversationsSetTopicArguments[];
        invitesToChannels: slack.ConversationsInviteArguments[];
        archivedChannels: slack.ConversationsCloseArguments[];
        sentMessages: slack.ChatPostMessageArguments[];
        lookedUpUsers: slack.UsersLookupByEmailArguments[];
    };

    constructor() {
        this.snapshot = {
            createdChannels: new Array<slack.ConversationsCreateArguments>(),
            setChannelTopics: new Array<slack.ConversationsSetTopicArguments>(),
            invitesToChannels: new Array<slack.ConversationsInviteArguments>(),
            archivedChannels: new Array<slack.ConversationsCloseArguments>(),
            sentMessages: new Array<slack.ChatPostMessageArguments>(),
            lookedUpUsers: new Array<slack.UsersLookupByEmailArguments>()
        };
    }

    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse> {
        this.snapshot.lookedUpUsers.push(options);
        return Promise.resolve({ ok: true, user: { id: options.email } });
    }

    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse> {
        this.snapshot.createdChannels.push(options);
        return Promise.resolve({ ok: true, channel: { id: channelId, name: options.name } });
    }

    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse> {
        this.snapshot.setChannelTopics.push(options);
        return Promise.resolve({ ok: true });
    }

    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve({ ok: true });
    }

    archiveChannel(options: slack.ConversationsArchiveArguments): Promise<slack.ConversationsArchiveResponse> {
        this.snapshot.archivedChannels.push(options);
        return Promise.resolve({ ok: true });
    }

    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true, channel: channelId });
    }
}
