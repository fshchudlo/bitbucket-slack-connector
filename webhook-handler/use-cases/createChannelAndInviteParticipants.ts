import {
    buildChannelName,
    formatUserName,
    getPullRequestDescriptionForSlack,
    slackLink,
    iconEmoji
} from "../slack-building-blocks";
import { SendMessageArguments, SlackAPIAdapter } from "../SlackAPIAdapter";
import { PullRequestBasicNotification } from "../../typings";

export async function createChannelAndInviteParticipants(payload: PullRequestBasicNotification, slackGateway: SlackAPIAdapter, createPrivateChannel: boolean) {
    const allParticipants = [payload.pullRequest.author.user].concat(payload.pullRequest.reviewers.map(r => r.user));
    const slackUserIds = await slackGateway.getSlackUserIds(allParticipants);

    const channelId = (
        await slackGateway.createChannel({
            name: buildChannelName(payload.pullRequest),
            isPrivate: createPrivateChannel
        })
    ).id;

    await slackGateway.setChannelTopic({ channelId, topic: buildChannelTopic(payload) });
    if (slackUserIds.length > 0) {
        await slackGateway.inviteToChannel({ channelId: channelId, users: slackUserIds, force: true });
    }
    await slackGateway.sendMessage(buildMessage(payload, channelId));
}

function buildMessage(payload: PullRequestBasicNotification, channelId: string): SendMessageArguments {
    const messageTitle = `The pull request was opened by ${formatUserName(payload.actor)}.`;
    const pleaseReviewText = `Please ${slackLink(payload.pullRequest.links.self[0].href, "review the PR")}`;
    const descriptionText = getPullRequestDescriptionForSlack(payload.pullRequest.description ?? payload.pullRequest.title);
    return {
        channel: channelId,
        iconEmoji: iconEmoji,
        text: messageTitle,
        attachments: [
            {
                text: descriptionText,
                color: "#0288D1"
            }, {
                text: pleaseReviewText
            }
        ]
    };
}

function buildChannelTopic({ pullRequest }: PullRequestBasicNotification) {
    const header = `${pullRequest.toRef.repository.project.key}/${pullRequest.toRef.repository.slug}:${pullRequest.toRef.displayId}`;
    let result = `:git: Pull request: ${slackLink(pullRequest.links.self[0].href, pullRequest.title)} | :git-branch: To branch: \`${header}\``;
    if (result.length > 250) {
        result = `:git: Pull request: ${pullRequest.title} | :git-branch: To branch: \`${header}\``;
    }
    return result;
}

