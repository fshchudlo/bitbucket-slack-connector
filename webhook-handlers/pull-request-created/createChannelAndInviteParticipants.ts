import buildChannelName from "../buildChannelName";

function buildChannelTopic({ pullRequest }: PullRequestCreatedPayload) {
    const header = `${pullRequest.toRef.repository.project.key}/${pullRequest.toRef.repository.slug}:${pullRequest.toRef.displayId}`;
    let result = `:git: Pull request: *<${pullRequest.links.self[0]}|${pullRequest.title}>* | :git-branch: To branch: *${header}*`;
    if (result.length > 250) {
        result = `:git: Pull request: *${pullRequest.title}* | :git-branch: To branch: *${header}*`;
    }
    return result;
}

export async function createChannelAndInviteParticipants(payload: PullRequestCreatedPayload, slackGateway: SlackGateway) {
    const pullRequest = payload.pullRequest;
    const channelName = buildChannelName(pullRequest.toRef.repository.project.key, pullRequest.toRef.repository.slug, pullRequest.id);

    const slackUserRequests = [pullRequest.author.user.emailAddress].concat(pullRequest.reviewers.map(r => r.user.emailAddress)).map(
        async email =>
            await slackGateway.lookupUserByEmail({
                email: email,
            })
    );

    const slackUserIds = [...new Set((await Promise.all(slackUserRequests)).map(r => r.user.id))];

    // Create a new Slack channel
    const channelId = (
        await slackGateway.createChannel({
            name: channelName,
            is_private: false,
        })
    ).channel.id;

    await slackGateway.setChannelTopic({
        channel: channelId,
        topic: buildChannelTopic(payload),
    });

    await slackGateway.inviteToChannel({
        channel: channelId,
        users: slackUserIds.join(","),
        force: true,
    });

    await slackGateway.sendMessage({
        channel: channelId,
        text: `The pull request was opened by ${pullRequest.author.user.name}. Please <${pullRequest.links.self[0]}|review the PR>`,
    });

    return channelId;
}
