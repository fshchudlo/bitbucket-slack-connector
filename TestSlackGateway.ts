import * as slack from "@slack/web-api";
import { SlackGateway } from "./contracts";

export class TestSlackGateway implements SlackGateway {
    snapshot: {
        createdChannels: slack.ConversationsCreateArguments[];
        setChannelTopics: slack.ConversationsSetTopicArguments[];
        invitesToChannels: slack.ConversationsInviteArguments[];
        closedChannels: slack.ConversationsCloseArguments[];
        sentMessages: slack.ChatPostMessageArguments[];
        lookedUpUsers: slack.UsersLookupByEmailArguments[];
    };
    constructor() {
        this.snapshot = {
            createdChannels: new Array<slack.ConversationsCreateArguments>(),
            setChannelTopics: new Array<slack.ConversationsSetTopicArguments>(),
            invitesToChannels: new Array<slack.ConversationsInviteArguments>(),
            closedChannels: new Array<slack.ConversationsCloseArguments>(),
            sentMessages: new Array<slack.ChatPostMessageArguments>(),
            lookedUpUsers: new Array<slack.UsersLookupByEmailArguments>(),
        };
    }
    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse> {
        this.snapshot.lookedUpUsers.push(options);
        return Promise.resolve({ ok: true, user: { id: options.email } });
    }
    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse> {
        this.snapshot.createdChannels.push(options);
        return Promise.resolve({ ok: true, channel: { id: "12345", name: options.name } });
    }
    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse> {
        this.snapshot.setChannelTopics.push(options);
        return Promise.resolve({ ok: true });
    }
    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve({ ok: true });
    }
    closeChannel(options: slack.ConversationsCloseArguments): Promise<slack.ConversationsCloseResponse> {
        this.snapshot.closedChannels.push(options);
        return Promise.resolve({ ok: true });
    }
    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true });
    }
}
