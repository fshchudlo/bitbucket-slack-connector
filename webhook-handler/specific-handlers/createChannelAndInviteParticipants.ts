import buildChannelName from "../helper-functions/buildChannelName";
import { slackLink, slackSection } from "../slack-building-blocks";
import { SlackGateway } from "../gateways/SlackGateway";
import getPullRequestDescriptionForSlack from "../helper-functions/getPullRequestDescriptionForSlack";
import { formatUserName } from "../slack-building-blocks/formatUserName";

function buildChannelTopic({ pullRequest }: PullRequestBasicNotification) {
    const header = `${pullRequest.toRef.repository.project.key}/${pullRequest.toRef.repository.slug}:${pullRequest.toRef.displayId}`;
    let result = `:git: Pull request: ${slackLink(pullRequest.links.self[0].href, pullRequest.title)} | :git-branch: To branch: \`${header}\``;
    if (result.length > 250) {
        result = `:git: Pull request: ${pullRequest.title} | :git-branch: To branch: \`${header}\``;
    }
    return result;
}

export async function createChannelAndInviteParticipants(payload: PullRequestBasicNotification, slackGateway: SlackGateway) {
    const pullRequest = payload.pullRequest;
    const channelName = buildChannelName(pullRequest.toRef.repository.project.key, pullRequest.toRef.repository.slug, pullRequest.id);
    const allParticipants = [pullRequest.author.user].concat(pullRequest.reviewers.map(r => r.user));
    const slackUserIds = await slackGateway.getSlackUserIds(allParticipants);

    const channelId = (
        await slackGateway.createChannel({
            name: channelName,
            is_private: true
        })
    ).channel.id;

    await slackGateway.setChannelTopic({
        channel: channelId,
        topic: buildChannelTopic(payload)
    });

    await slackGateway.inviteToChannel({
        channel: channelId,
        users: slackUserIds.join(","),
        force: true
    });

    const messageTitle = `The pull request was opened by ${formatUserName(payload.actor)}.`;
    const pleaseReviewText = `Please ${slackLink(pullRequest.links.self[0].href, "review the PR")}`;
    const descriptionText = getPullRequestDescriptionForSlack(pullRequest.description ?? pullRequest.title);

    await slackGateway.sendMessage({
        channel: channelId,
        text: `${messageTitle} ${pleaseReviewText}`,
        blocks: [
            slackSection(messageTitle),
            slackSection(descriptionText),
            slackSection(pleaseReviewText)]
    });
}

