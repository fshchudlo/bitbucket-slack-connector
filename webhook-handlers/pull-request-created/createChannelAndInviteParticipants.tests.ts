// mainApp.test.ts
import { TestSlackGateway } from "../../TestSlackGateway";
import { PullRequestCreatedPayload } from "../../contracts";
import { createChannelAndInviteParticipants } from "./createChannelAndInviteParticipants";

let payload: PullRequestCreatedPayload = {
    eventKey: "pr:opened",
    date: "2017-09-19T09:58:11+1000",
    pullRequest: {
        id: 1,
        title: "Quite a long and comprehensive description to get channel title longer than 250 symbols which is a maximum length for the Slack",
        toRef: {
            displayId: "master",
            repository: {
                slug: "test-repository",
                project: {
                    key: "TEST-PROJ",
                    name: "Test project",
                },
            },
        },
        author: {
            user: {
                name: "test.author",
                emailAddress: "test.author@test.com",
            },
        },
        reviewers: [
            {
                user: {
                    name: "test.reviewer1",
                    emailAddress: "test.reviewer1@test.com",
                    },
            },
        ],
        links: {
            self: ["https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview"],
        },
    },
};

describe("createChannelAndInviteParticipants", () => {
    it("Should create channel, set topic and invite author and reviewers", async () => {
        const testee = new TestSlackGateway();

        await createChannelAndInviteParticipants(payload, testee);

        expect(testee.snapshot).toMatchSnapshot();
    });
});
