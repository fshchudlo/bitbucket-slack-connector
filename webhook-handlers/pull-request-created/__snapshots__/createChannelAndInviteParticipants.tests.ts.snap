// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createChannelAndInviteParticipants Should create channel, set topic and invite author and reviewers 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [
    {
      "is_private": false,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "invitesToChannels": [
    {
      "channel": "12345",
      "force": true,
      "users": "test.author@test.com,test.reviewer1@test.com",
    },
  ],
  "lookedUpUsers": [
    {
      "email": "test.author@test.com",
    },
    {
      "email": "test.reviewer1@test.com",
    },
  ],
  "sentMessages": [
    {
      "channel": "12345",
      "text": "The pull request was opened by test.author. Please <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|review the PR>",
    },
  ],
  "setChannelTopics": [
    {
      "channel": "12345",
      "topic": ":git: Pull request: *Quite a long and comprehensive description to get channel title longer than 250 symbols which is a maximum length for the Slack* | :git-branch: To branch: *TEST-PROJ/test-repository:master*",
    },
  ],
}
`;
